// ng2-image-lazy-load/src/app/services/web-worker.service.stub.js
"use strict";
var core_1 = require('@angular/core');
var WebWorkerStub = (function () {
    function WebWorkerStub() {
        this.activeWorkers = [];
    }
    //public workersCalled: number = 0;
    WebWorkerStub.prototype.load = function (config, msgFn, errorFn) {
        var id = Math.floor(Math.random() * 1000000000000);
        this.activeWorkers.push({
            id: id,
            config: config
        });
        setTimeout(function () {
            if (msgFn) {
                msgFn({ data: 'success' });
            }
            else if (errorFn) {
                errorFn({ data: 'ERROR' });
            }
        });
        //this.workersCalled++;
        return id;
    };
    WebWorkerStub.prototype.terminate = function (id) {
        var index = this.activeWorkers.findIndex(function (item) { return item.id === id; });
        if (index > -1) {
            this.activeWorkers.splice(index, 1);
        }
    };
    WebWorkerStub.supported = true;
    WebWorkerStub.workerUrl = 'assets/js/xhrWorker.js';
    WebWorkerStub = __decorate([
        core_1.Injectable(), 
        __metadata('design:paramtypes', [])
    ], WebWorkerStub);
    return WebWorkerStub;
}());
exports.WebWorkerStub = WebWorkerStub;
